File: 22.generate-parentheses.py
#
# @lc app=leetcode id=22 lang=python3
#
# [22] Generate Parentheses
#

# @lc code=start
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def gen_par_helper(left : int, right: int) -> list:
            if left < 0 or right < 0:
                return []
            if left == 0 and right == 1:
                return [")"]
            if left == 0 and right == 0:
                return []
            if left > right :
                return []
            if left == right:
                return ["(" + x for x in gen_par_helper(left - 1, right)]
            return [")" + x for x in gen_par_helper(left, right - 1)] + \
                ["(" + x for x in gen_par_helper(left - 1, right)]
        
        if n <= 0:
            return []
        return gen_par_helper(n, n)













#         def gen_parenthesis_helper(left_num : int, right_num : int) -> List[str]:
#             if left_num == 0:
#                 if right_num == 0:
#                     return [""]
#                 return [")" + x for x in gen_parenthesis_helper(0, right_num - 1)]
#             if left_num == right_num:
#                 return ["(" + x for x in gen_parenthesis_helper(left_num - 1, right_num )]
#             else:
#                 return [")" + x for x in gen_parenthesis_helper(left_num, right_num - 1)] + \
#                     ["(" + x for x in gen_parenthesis_helper(left_num - 1, right_num )]
#         if n <= 0:
#             return []
#         return gen_parenthesis_helper(n, n)




# def generateParenthesis(n: int) -> list:
#     def gen_parenthesis_helper(left_num : int, right_num : int) -> list:
#         if left_num == 0:
#             if right_num == 0:
#                 return [""]
#             return [")" + x for x in gen_parenthesis_helper(0, right_num - 1)]

#         if left_num == right_num:
#             return ["(" + x for x in gen_parenthesis_helper(left_num - 1, right_num )]
#         else:
#             return [")" + x for x in gen_parenthesis_helper(left_num, right_num - 1)] + \
#                 ["(" + x for x in gen_parenthesis_helper(left_num - 1, right_num )]


#     if n <= 0:
#         return []
#     return gen_parenthesis_helper(n, n)


# if __name__ == '__main__':
#     n = 10
#     res = generateParenthesis(n)
#     print(res)
#     print(len(res))

# @lc code=end

File: 22.generate-parentheses.cpp
/*
 * @lc app=leetcode id=22 lang=cpp
 *
 * [22] Generate Parentheses
 */

// @lc code=start



vector<string> generate_par_helper(int left, int right){
    vector<string> res;
    if (left == 0){
        string s(right, ')');
        res.push_back(s);
        return res;
    }
    if (left == right){
        vector<string> next_list = generate_par_helper(left - 1, right);
        for (int i = 0; i < next_list.size(); ++i){
            res.push_back("(" + next_list[i]);
        }
    }
    else{
        vector<string> next_list_1 = generate_par_helper(left -1, right);
        vector<string> next_list_2 = generate_par_helper(left, right - 1);
        for (int i = 0; i < next_list_1.size(); ++i){
            res.push_back("(" + next_list_1[i]);
        }
        for (int i = 0; i < next_list_2.size(); ++i){
            res.push_back(")" + next_list_2[i]);
        }
    }
    return res;
}
class Solution {
public:
    vector<string> generateParenthesis(int n) {


        if (n ==0){
            return {};
        }
        return generate_par_helper(n, n);
    }
};
// @lc code=end

